cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -87,6 +89,203 @@ Status GetListElementIndex(const ExecValue& value, T* out) {
   return Status::OK();
 }
 
+template <typename T>
+std::string ToString(const std::optional<T>& o) {
+  return o.has_value() ? std::to_string(*o) : "(nullopt)";
+}
+
+template <typename Type, typename IndexType>
+struct ListSlice {
+  using offset_type = typename Type::offset_type;
+
+  static Status Exec(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
+    const auto opts = OptionsWrapper<ListSliceOptions>::Get(ctx);
+
+    // Invariants
+    if (!opts.stop.has_value()) {
+      // TODO(ARROW-18280): Support slicing to arbitrary end
+      // For variable size list, this would be the largest difference in offsets
+      // For fixed size list, this would be the fixed size.
+      return Status::NotImplemented(
+          "Slicing to end not yet implemented, please set `stop` parameter.");
+    }
+    if (opts.start < 0 || opts.start >= opts.stop.value()) {
+      // TODO(ARROW-18281): support start == stop which should give empty lists
+      return Status::Invalid("`start`(", opts.start,
+                             ") should be greater than 0 and smaller than `stop`(",
+                             ToString(opts.stop), ")");
+    }
+    if (opts.step != 1) {
+      // TODO(ARROW-18282): support step in slicing
+      return Status::NotImplemented(
+          "Setting `step` to anything other than 1 is not supported; got step=",
+          opts.step);
+    }
+
+    const ArraySpan& list_array = batch[0].array;
+    const Type* list_type = checked_cast<const Type*>(list_array.type);
+    const auto field_name = list_type->field(0)->name();
+    const auto value_type = list_type->field(0)->WithName(field_name);
```

Comment:
I am not sure the `WithName` is actually needed here? Since you are getting the field of `list_type` as well, it should already have that name?

So this could be simplified to:

```suggestion
    const auto value_type = list_type->field(0);
```

?